File: example/DFS/DFS.js
const graph = [[], [2, 3, 8], [1, 7], [1, 4, 5], [3, 5], [3, 4], [7], [2, 6, 8], [1, 7]];
// 각 인덱스에 연결된 노드
const visited = Array(9).fill(false);
let result = "";

dfs(graph, 1, visited);
console.log(result);

function dfs(graph, v, visited) {
  visited[v] = true;
  result += `${v} `;
  for (i of graph[v]) {
    if (!visited[i]) dfs(graph, i, visited);
    // 그 노드가 방문하지 않은 노드인면 dfs 재귀
  }
}

File: java/src/algorithm/DFS/DFS.java
package algorithm.DFS;

import java.util.ArrayList;
import java.util.List;

// 시간 복잡도: O(V + E) - V: 정점의 수, E: 간선의 수
// 문제 유형
// 1. 영역 탐색, 연결 요소 세기,
// 2. 모든 경로 탐색, 백트래킹
// 3. 영역 채우기
// 4. 재귀적 성질이 있는 문제
public class DFS {

    static boolean[] visited = new boolean[9];
    static List<List<Integer>> graph = new ArrayList<>();
    static int[][] edges = {
            {1, 2},
            {1, 3},
            {1, 8},
            {2, 7},
            {3, 4},
            {3, 5},
            {4, 5},
            {6, 7},
            {7, 8}
    };

    public static void main(String[] args) {
        for (int i = 0; i < 9; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int a = edge[0];
            int b = edge[1];

            graph.get(a).add(b);
        }

        dfs(1); // 1부터 시작
    }

    public static void dfs(int v) {
        visited[v] = true;
        System.out.print(v + " ");

        for (int i : graph.get(v)) {
            if (!visited[i]) {
                dfs(i);
            }
        }
    }
}
